### PR TITLE
Recalculate plot on unit change

### PR DESCRIPTION
**Pull request by ni-m**
This PR adds the functionality to update the plot after a unit change is detected.

https://github.com/linescaleGUI/linescaleGUI/blob/f7258ae9e37975dda0561a254b9a175528592414/src/gui/plotWidget.cpp#L266-L299

### CODE REVIEW

File: src/gui/plotWidget.cpp [hunk 1]
```diff
@@ -262,3 +262,38 @@ void Plot::clearSelection() {
...
+    }
+
+    double factorBack = 1;
+    double factorForward = 1;
+
+    factorBack = (current == UnitValue::LBF) ? 1 / factorKnToLbf : factorBack;
+    factorBack = (current == UnitValue::KGF) ? 1 / factorKnToKgf : factorBack;
+
+    factorForward = (next == UnitValue::LBF) ? factorKnToLbf : factorForward;
+    factorForward = (next == UnitValue::KGF) ? factorKnToKgf : factorForward;
```
Comment (N3xed) [inline src/gui/plotWidget.cpp:279]: Possibly use a `switch case`.

File: src/gui/plotWidget.cpp [hunk 2]
```diff
@@ -262,3 +262,38 @@ void Plot::clearSelection() {
         graph->setSelection(QCPDataSelection{});
     }
 }
+
+void Plot::convertToNewUnit(UnitValue current, UnitValue next) {
+    // Check for a valid unit
+    if(current != UnitValue::KN && current != UnitValue::KGF && current != UnitValue::LBF) {
+        return;
+    }
```
Comment (N3xed) [inline src/gui/plotWidget.cpp:270]: This shouldn't need to be checked. A better check would be if the unit to be converted too is already the current unit.

File: src/gui/plotWidget.cpp [hunk 3]
```diff
@@ -262,3 +262,38 @@ void Plot::clearSelection() {
         graph->setSelection(QCPDataSelection{});
     }
 }
+
+void Plot::convertToNewUnit(UnitValue current, UnitValue next) {
```
Comment (N3xed) [inline src/gui/plotWidget.cpp:266]: Why have a `current` argument and a `Plot::lastUnit` at the same time?
I'd suggest using the `Plot::lastUnit` property.

Comment (ni-m) [reply src/gui/plotWidget.cpp]: Removed current, changed to Plot::lastUnit

File: src/gui/plotWidget.cpp (hunk 2 above)
Comment (ni-m) [reply src/gui/plotWidget.cpp]: Check isn't needed anymore.
The check wether the unit has changed is already done before the function call.
https://github.com/linescaleGUI/linescaleGUI/blob/a05e133a64655be318c9eb4186151b2e73f7c3f2/src/gui/plotWidget.h#L84-L90

File: src/gui/plotWidget.cpp (hunk 1 above)
Comment (ni-m) [reply src/gui/plotWidget.cpp]: Changed to switch case

File: src/gui/plotWidget.h [hunk 4]
```diff
@@ -134,30 +142,50 @@ class Plot : public QWidget {
...
+     * @brief Clear the current selection inside the plot.
      */
     void clearSelection();
 
+    /**
+     * @brief Update the plot after a unit change was detected
+     *
+     * Calculation is done in two steps. First from the current unit to kN
+     * and afterwards from kN to the next unit. This reduces the amount of
+     * different factors.
```
Comment (N3xed) [inline src/gui/plotWidget.h:162]: That's not really true, the calculation is done in one step. This explanation makes it seem that this operation takes a long time. Also, I think it's better to let this be an implementation detail.
```suggestion
     * To go from the current unit to the `next`, all values in the plot are multiplied by a conversion factor.
```

File: src/gui/plotWidget.h [hunk 5]
```diff
@@ -134,30 +142,50 @@ class Plot : public QWidget {
...
+     */
+    void convertToNewUnit(UnitValue next);
+
    private:
     QCustomPlot* customPlot;
     double minValue = 0.0, maxValue = 0.0;
     double lastTime = 0.0;
     bool hadNewData = false;
 
+    UnitValue lastUnit;
```
Comment (N3xed) [inline src/gui/plotWidget.h:177]: Maybe change to `currentUnit`.

File: src/gui/plotWidget.h [hunk 6]
```diff
@@ -77,11 +77,15 @@ class Plot : public QWidget {
...
      *
-     * @todo Handle different `Sample::unitValue`s.
+     * If the current unit differs from the previous, the existing plot will be recalculated.
      *
      * @param sample The sample to add.
      */
     inline void addConsecutiveSample(const Sample& sample) {
+        if (lastUnit != sample.unitValue) {
+            convertToNewUnit(sample.unitValue);
+            lastUnit = sample.unitValue;
```
Comment (N3xed) [inline src/gui/plotWidget.h:87]: I'd move this into the `convertToNewUnit` function.

**Review by N3xed — COMMENTED** [review]
Some small nitpicks, otherwise looks good.

**Review by N3xed — DISMISSED** [review]

**Review by N3xed — APPROVED** [review]

**Pull request merged**